### PR TITLE
[inductor] Do variance calculation in opmath type

### LINKED ISSUE #114903: [inductor] `torch.var` unexpectedly returns NaN values on A16 and A2
### 🐛 Describe the bug

Compiling a minimal module that only does `torch.var` produces code that unexpectedly yields NaN values when executed on A16 and A2. Interestingly, the issue doesn't seem to reproduce on all SM 8.6 devices (such as A40).

```
import torch

import torch._dynamo as dynamo

torch._dynamo.config.repro_after="aot"

class Model(torch.nn.Module):
    def forward(self, x):
        return torch.var(x)

model = Model()

def test_backend_error():
    x = torch.rand((10, 3, 352, 352), dtype=torch.float16, device='cuda')
    y = model(x)
    assert not torch.isnan(y)

compiled_test_backend_error = torch.compile(test_backend_error, backend="inductor")
compiled_test_backend_error()
```

### Versions

Tested on a fresh (as of a few hours ago) source build.

cc @ptrblck @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115181

Fixes #114903

Previously large split variance reductions stored the intermediates as float16
precision, which may lead to overflow as the intermediate result is
unnormalized.

In #114903 we see two different `num_split` decisions made based on the
hardware capabilities, one of which has large enough intermediates to cause
overflows.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -1096,6 +1096,13 @@ def fn(a):
         self.common(fn, ((torch.rand((10, 3, 352, 352), dtype=torch.float32),)))
         self.common(fn, ((torch.rand((14923), dtype=torch.float32),)))
 
+    def test_multilayer_var_lowp(self):
+        def fn(a):
+            return torch.var(a)
+
+        self.common(fn, ((torch.rand((10, 3, 352, 352), dtype=torch.float16),)))
```
Comment (shunting314) [inline test/inductor/test_torchinductor.py:1103]: Does this test repro the issue? Or should we force different number of multi-processsor to simulate the GPU that the author encounter the failure?

Comment (peterbell10) [reply test/inductor/test_torchinductor.py]: I've bumped the input size and confirmed it now fails on an A100 machine.

**Review by shunting314 — APPROVED** [review]

**Pull request closed without merging**